### PR TITLE
DATAINT-1642: Entity Relationship Information Discovery for MSSQL Connections

### PR DESCRIPTION
**Pull request by TerryNue**
resolves [DATAINT-1642](https://aunalytics.atlassian.net/browse/DATAINT-1642?atlOrigin=eyJpIjoiNWZmMWUyNzlkMzgwNGI2YmJmOWYwYjFlNDM2ZmQ5MjEiLCJwIjoiaiJ9)

-identifies and retrieves foreign key relationships between database schemas

updates: 9/7/23
-adjusts query to account for multi-part foreign key relationships
-differentiate between single and multi-part foreign keys, labeling them as 'FOREIGN KEY' and 'MULTI-PART FOREIGN KEY'

update: 9/15/23
-fixes parsing of schema names to store correct value in schemaNames slice

[DATAINT-1642]: https://aunalytics.atlassian.net/browse/DATAINT-1642?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: internal/server.go [hunk 1]
```diff
@@ -481,6 +481,100 @@ func (s *Server) DiscoverSchemas(ctx context.Context, req *pub.DiscoverSchemasRe
...
+		return nil, errors.New("no schemas provided to relate")
+	}
+
+	placeholders := make([]string, len(schemaNames))
+	for i := range schemaNames {
+		placeholders[i] = fmt.Sprintf("@p%d", i+1)
+	}
+	whereClause := fmt.Sprintf("WHERE OBJECT_SCHEMA_NAME(fk.parent_object_id) IN (%s)", strings.Join(placeholders, ","))
+
+	query := `
```
Comment (roehlerw) [inline internal/server.go:509]: this doesn't account for potentially multi-part foreign key relationships

**Review by roehlerw — CHANGES_REQUESTED** [review]

**Review by roehlerw — CHANGES_REQUESTED** [review]

**Review by roehlerw — APPROVED** [review]

File: internal/server.go [hunk 2]
```diff
@@ -481,6 +481,77 @@ func (s *Server) DiscoverSchemas(ctx context.Context, req *pub.DiscoverSchemasRe
 	}, err
 }
 
+func (s *Server) DiscoverRelatedEntities(ctx context.Context, req *pub.DiscoverRelatedEntitiesRequest) (*pub.DiscoverRelatedEntitiesResponse, error) {
```
Comment (roehlerw) [inline internal/server.go:484]: Needs to be modified to loop through the input schemas on the request and only return related entities to the ones requested.

**Review by roehlerw — CHANGES_REQUESTED** [review]

**Review by g1hanna — APPROVED** [review]
Looks nice! A comment from Wyatt needs to be marked as resolved (it looks like you've applied the feedback). The SQL script looks great, too!

File: internal/server.go [hunk 3]
```diff
@@ -481,6 +481,104 @@ func (s *Server) DiscoverSchemas(ctx context.Context, req *pub.DiscoverSchemasRe
...
+
+	db := s.session.DB
+	if db == nil {
+		return nil, errors.New("database connection not available")
+	}
+
+	var schemaNames []string
+	for _, schema := range req.GetToRelate() {
+		parts := strings.Split(schema.GetName(), ".")
```
Comment (roehlerw) [inline internal/server.go:493]: This should be based on the Id and not the Name as the user can change the Schema Name but not the Id.

You should also use the function DecomposeSafeName to decode the Schema Id into parts

**Review by roehlerw — CHANGES_REQUESTED** [review]

**Review by roehlerw — APPROVED** [review]

**Review by uttam028 — APPROVED** [review]

**Review by g1hanna — APPROVED** [review]

**Pull request merged**